packages/asset-swapper/src/utils/quote_requestor.ts

```diff
@@ -0,0 +1,100 @@
+import { assetDataUtils, SignedOrder } from '@0x/order-utils';
+import { ERC20AssetData } from '@0x/types';
+import { BigNumber, logUtils } from '@0x/utils';
+import Axios, { AxiosResponse } from 'axios';
+import * as _ from 'lodash';
+
+import { constants } from '../constants';
+import { MarketOperation, RfqtFirmQuoteRequestOpts } from '../types';
+
+/**
+ * Request quotes from RFQ-T providers
+ */
+
+function getTokenAddressOrThrow(assetData: string): string {
+    const decodedAssetData = assetDataUtils.decodeAssetDataOrThrow(assetData);
+    if (decodedAssetData.hasOwnProperty('tokenAddress')) {
+        // type cast necessary here as decodeAssetDataOrThrow returns
+        // an AssetData object, which doesn't necessarily contain a
+        // token address.  (it could possibly be a StaticCallAssetData,
+        // which lacks an address.)  so we'll just assume it's a token
+        // here.  should be safe, with the enclosing guard condition
+        // and subsequent error.
+        // tslint:disable-next-line:no-unnecessary-type-assertion
+        return (decodedAssetData as ERC20AssetData).tokenAddress;
+    } else {
+        throw new Error(`Decoded asset data (${JSON.stringify(decodedAssetData)}) does not contain a token address`);
+    }
+}
+
+export class QuoteRequestor {
+    private readonly _rfqtMakerEndpoints: string[];
+
+    constructor(rfqtMakerEndpoints: string[]) {
+        this._rfqtMakerEndpoints = rfqtMakerEndpoints;
+    }
+
+    public async requestRfqtFirmQuotesAsync(
+        makerAssetData: string,
+        takerAssetData: string,
+        assetFillAmount: BigNumber,
+        marketOperation: MarketOperation,
+        takerApiKey: string,
+        takerAddress: string,
+        options?: Partial<RfqtFirmQuoteRequestOpts>,
+    ): Promise<SignedOrder[]> {
+        const { makerEndpointMaxResponseTimeMs } = _.merge({}, constants.DEFAULT_RFQT_FIRM_QUOTE_REQUEST_OPTS, options);
+
+        const buyToken = getTokenAddressOrThrow(makerAssetData);
+        const sellToken = getTokenAddressOrThrow(takerAssetData);
+
+        // create an array of promises for quote responses, using "undefined"
+        // as a placeholder for failed requests.
+
+        const responsesIfDefined: Array<void | AxiosResponse<SignedOrder>> = await Promise.all(
+            this._rfqtMakerEndpoints.map(async rfqtMakerEndpoint => {
+                try {
+                    return Axios.get<SignedOrder>(`${rfqtMakerEndpoint}/quote`, {
```

Comment:
Since we're not awaiting this it won't land in the try/catch right? We should await here to make sure it does lang as expected.

We could probably have a test on this behaviour that when RFQT HTTP request does blow up, the user still gets a quote

Reply:
We could also just do:
```
const response = await Axios.get<SignedOrder>...
return response.data
```
Then we won't need to `filter` and `map` iin the lines below, just filter for `undefined`.

Reply:
@steveklebanoff maybe it's too late but do you have any thoughts on an idempotency of this request to the MM? Let's say we retry due to a network blip, we'd retry with the same `id` so the MM (if they did receive the first request, but we failed to get the response) would understand we're just asking for the same thing again, rather than a second quote for the taker.

This idempotency id could be in the params or as a header.

Reply:
I think this is a cool idea for a follow-up iteration of RFQT, but out of scope for this first MVP iteration.

I added this idea to our working "RFQT Spec v3 Potential Changes" document 👍

Reply:
@dekz I committed https://github.com/0xProject/0x-monorepo/commit/026bd5802cf774654debd6549a220fc9d861c482 to address the try/catch issue you pointed out.  Is that what you had in mind?

Reply:
@dekz I committed https://github.com/0xProject/0x-monorepo/pull/2541/commits/0cb5e4553b4de9cb7f9503bce4410d1d366a0dd0 to address the try/catch issue you pointed out. Is that what you had in mind?